### ISSUE TITLE
Switch back to released PyInstaller

### ISSUE THREAD

**Issue by toofar** [qutebrowser/qutebrowser#7806]
In https://github.com/qutebrowser/qutebrowser/pull/7803 we are changing our PyInstaller dependency to pull in a version from their develop branch, rather than a released version. We had to put in place a workaround to get that working. This issue is to track reverting those workarounds once a new major version (6.0+) is out.

Currently these commits are the ones to revert, they are around recompiling the bootloader in CI (leaving them there probably wouldn't break anything, technically). Hopefully the commit IDs are still correct (check the PR linked above):

* 6042fccbf589 Re-compile PyInstaller bootloader in CI.
* 64af5f0e264f walk_module workaround for PyInstaller (ref https://github.com/qutebrowser/qutebrowser/pull/7803#discussion_r1305429801)
* ~dd86483065a8 Move PYINSTALLER_COMPILE_BOOTLOADER into build script~ (already reverted)
* ~8a327beb17db tox: disable pip cache for pyinstaller environment~ (already reverted)

<details>
<summary>
Already reverted commits:
</summary>

```patch
commit 8a327beb17dbe97cc8022549d652f72eb2996c97
Author: toofar <toofar@spalge.com>
Date:   Sun Jul 30 12:22:44 2023 +1200

    tox: disable pip cache for pyinstaller environment
    
    The build_release script which calls pyinstaller now sets
    PYINSTALLER_COMPILE_BOOTLOADER=true which is used when building the
    wheel. On ci the build_release script runs in a tox env which also
    installs pyinstaller (something about updating version info on windows).
    So by the time the pyinstaller venv is built there is already a cached
    pyinstaller wheel. This commit disables that cache so the wheel should
    always be rebuilt.
    
    It looks like we can't put arbitrary pip CLI args in `deps` anymore. And
    I couldn't get install_command to inherit from the parent env (I tried
    `{[testenv]install_command} --no-cache-dir`) so had to reproduce the
    command I observed with `tox -vv`.
    
    Alternatives:
    * set the env var in the tox config (is the manual build done via tox
      anyway? In that case there's no point in trying to move the env var
      into the script anyway)
    * set the env var in the workflow config (and pass through in tox) -
      doesn't help with the manual build case which is the reason for moving
      the env var into the script anyway
    
    This should be reverted once a pyinstaller 6 release is out.

diff --git a/tox.ini b/tox.ini
index 36814b8203e9..d80f1ce3259d 100644
--- a/tox.ini
+++ b/tox.ini
@@ -190,6 +190,8 @@ passenv =
     PYINSTALLER_COMPILE_BOOTLOADER
 setenv =
     qt5: PYINSTALLER_QT5=true
+install_command =
+    python -I -m pip install --no-cache-dir {opts} {packages}
 deps =
     -r{toxinidir}/requirements.txt
     -r{toxinidir}/misc/requirements/requirements-pyinstaller.txt

commit dd86483065a81726e3ce70ec9fab7389c12efeb4
Author: toofar <toofar@spalge.com>
Date:   Thu Jul 27 21:09:11 2023 +1200

    Move PYINSTALLER_COMPILE_BOOTLOADER into build script
    
    Previously this was being set in tox which means that if you run build_release
    manually, not via tox, it won't be set. So this commit moves the env var to be
    set in the build_release script when pyinstaller is being invoked.
    I'm not actually sure how the manual builds are done though, maybe they are
    done via tox too.
    
    This should be reverted once a pyinstaller 6 release is out.

diff --git a/scripts/dev/build_release.py b/scripts/dev/build_release.py
index f602cae43250..b2af109832b7 100755
--- a/scripts/dev/build_release.py
+++ b/scripts/dev/build_release.py
@@ -81,6 +81,7 @@ def call_tox(
     env['PATH'] = os.environ['PATH'] + os.pathsep + str(python.parent)
     if debug:
         env['PYINSTALLER_DEBUG'] = '1'
+    env['PYINSTALLER_COMPILE_BOOTLOADER'] = 'true'
     subprocess.run(
         [sys.executable, '-m', 'tox', '-vv', '-e', toxenv, *args],
         env=env, check=True)
diff --git a/tox.ini b/tox.ini
index af59910f2080..36814b8203e9 100644
--- a/tox.ini
+++ b/tox.ini
@@ -269,7 +269,6 @@ passenv = *
 # Override default PyQt6 from [testenv]
 setenv =
     qt5: QUTE_QT_WRAPPER=PyQt5
-    PYINSTALLER_COMPILE_BOOTLOADER=true
 usedevelop = true
 deps =
     -r{toxinidir}/requirements.txt

```
</details>

**Comment by toofar** [comment]
Looks like the new PyInstaller release is out now: https://github.com/pyinstaller/pyinstaller/releases/tag/v6.0.0
I'll try to make time to upgrade it and go through the changelog next week.

**Comment by toofar** [comment]
We should be back to using the latest upstream pyinstaller with no workarounds now. I tested with the builds from this run: https://github.com/qutebrowser/qutebrowser/actions/runs/6671608688
Although I should figure out how to remove that action now, the normal nightly build will run later today and give us some new ones anyway.

b4bbf3252f7678caf9443fa963cd465b58d8181c
6e534c74e7e36429c04c304d819bcec815007c06
1fa678db61f2619ed14adff737ba07a783680ed3
ab7b07941108f474da31716a2964adf081c55049

The only other thing I saw in the changelog that might be relevant to us (beyond the onedir/onefile constructs which I don't understand) is the ability to pass CLI args to the spec file now: https://pyinstaller.org/en/v6.1.0/spec-files.html#adding-parameters-to-spec-files
We are currently passing a debug flag via an env var, and that could probably be changed to be via a CLI arg now. But I don't plan on touching that as testing it would require waiting for builds to run, and I'm not in the mood.

Additionally I did plan to make some changes around debug builds and smoke test but I'm just going to leave that for now. The problem I was seeing was that for the nightly action `--debug` jobs if the applications fails the smoke test we don't see the logs because they are captured. One option that would work fine is to just not capture the output (and also warn on non-zero exit code) if the job has `--debug` passed. But I feel like there might be a bit more refactoring possible there to make the situation a bit more clear:

* I think the point of the debug jobs is to produce a binary for download even if the smoke test fails, eg to do the build but skip the smoke test, so why aren't we doing that entirely?
* It does currently skip the failure path from the smoke tests, but it still runs them
* The `--debug` also gets passed into the pyinstaller EXE class in the spec file which results in [a progress bar](https://github.com/pyinstaller/pyinstaller/blob/develop/PyInstaller/building/api.py#L364)
* And it makes it so the built artifact always runs as if `--debug` was passed (eg print debug logs etc)
* In the smoke test, for the non `--debug` jobs, if they fail there is a second smoke test run that passes the `--debug` arg to qutebrowser (not pyinstaller) to get extra debug logs to try to surface more information about why the smoke test failed. This is useful but confused the heck out of me because now there are two different `--debug` flags in play

It kinda seems straightforward when I explain it like that but I found `--debug` being used for two different purposes (gimmie an artifact even if the smoke test fails vs get extra logs from the smoke test) in the same (or closely related) bits of code/infra to be confusing. Possibly the debug jobs could be more clearly named.
Also do they need to be separate jobs? Can we upload artifacts and fail the job? Change the name of them to make clear they are broken? Just seems like some duplication of effort. Like you probably only want the failing artifacts if the smoke test fails anyway, so they don't need to be running every night.

Anyway, the whole thing is working well enough and I didn't want to spend any more energy on it. I'm sure we'll have another crack at it if we run into it again :)
